docs/developing/app-review-guide.md

```diff
@@ -0,0 +1,84 @@
+
+The reviewer may look at the changelog and commit history to narrow down
+what new features or changes should be tested, and may reach out to the
+developer with suggestions or feedback. However, in many cases, the app reviewer
+will simply approve the update after verifying its functionality and notify the
+developer that it has been approved.
+
+## Approval-blocking issues
+
+The Sandstorm App Market is a fairly permissive environment. While you may
+receive feedback about performance, functionality, or usability, we largely
+require only that your app works and that it does not lose data.
+
+We will reject an app which crashes immediately and does not function as
+advertised, for example. This can often occur when the package is missing files
+that were present in the dev environment and wasn't adequately tested prior to
+submission.
+
+The latter concern often occurs when a pure client app is submitted up that
+implements localStorage as its storage medium, which Sandstorm does not effectively
+support. This can often be tested by storing some information in a grain, allowing
+Sandstorm to shut down that grain, and then opening it again and seeing if things
+persisted that an ordinary user would expect to persist between sessions.
+
+Finally, the App Market may reject apps which present security or privacy issues
+that fall outside of what a Sandstorm user would expect when using the app. This
+can mean it requests overly-broad permissions when making Powerbox requests or
+uses client-side loading to load scripts which share unexpected information outside
+of the Sandstorm sandbox. For example, we may block approval on an app which attempts
+to load an analytics script from an outside server, or does not make it suitably
+clear to a user why it needs certain outside network access prior to requesting it.
+
+**Note:** Client-side loading is a known gap in Sandstorm's security model that we
+intend to close. Not only is misuse of it a security or privacy issue, it's use may
+lead to apps that break when Sandstorm closes this gap.
+
+## Commercial or proprietary apps
+
+The Sandstorm App Market supports both commercial and proprietary apps, however
+official support for app purchase and licensing is not currently available. For an
+app that is merely closed source, but free, this may not pose an issue. For apps
+which require some form of licensing to function, we expect at minimum that the
+app description clearly and prominently specifies the licensing requirements and
+activation method. Sandstorm apps should be as self-sufficient as possible, so a
+license key method using some sort of signing that does not need to call home is ideal,
+however, a check with a licensing server may be acceptable if it is narrow in scope,
+and utilizes the Powerbox to get the user's consent for their server to contact
+yours. We would also encourage the app to have at least a minimal level of functionality
```

Comment:
I'm not sure I want to be even this encouraging re: licensing stuff that phones home; I think we should default to treating that as an overly broad permission.

I feel like if we're at the point where people are setting up licensing servers for Sandstorm apps folks have lost the plot on _most_ of sandstorm's goals. The developer is still needing to run infra for their users, the users don't have control over their use of the app... what's the point?

Reply:
So, coming from the enterprise IT world: Most on-premise paid software these days has some sort of licensing check. The benefits are still apparent, as it should work locally even when offline (I've dealt with one particularly bad software implementation where server issues with their licensing infrastructure shut down the whole app on-prem immediately... ugh) and very crucially, that any customer/business/proprietary data remains on your server and storage instead of the software developer's which is important for security and compliance.

I don't feel like the language I chose here encourages a phone home, because I encourage an offline method explicitly (Sandstorm for Work's feature key strategy is really what would be best here), but many apps that aren't written-for-Sandstorm that might want to run on Sandstorm may be designed with a call home, and (personal opinion) I feel we should not explicitly prevent that provided the user is choosing to authorize it via Powerbox and accurately informed of why the request is needed. I would agree that natively-written-for-Sandstorm apps should definitely not be written with a requirement on outside network access on principle, and that a feature key type strategy is where they should go.

I think the benefits for Sandstorm remain mostly intact: The app operates in a limited sandbox, which is especially important for proprietary apps, and the data remains locally stored and under the user's control. Probably the "Sandstorm has unlimited resources" solution would be for Sandstorm to have a specific type of Powerbox request for licensing/purchases that regulates what information is transmitted and received, but as you said, it's not a good idea for us to invest resources in developing tools for proprietary software, especially as a community project not in a position to collect a cut of it. Such as we aren't going to do that, we probably should accept that some people might want to implement their own.

Of course, we could elect to decide that if you want to do this, you should distribute your SPK outside of the App Market, but I feel that it would be unfortunate if we had major business tools that ran on our platform but people didn't know it was possible because we didn't have them in the app market.

"We would also encourage" is kinda meant to be separate, by the way, regardless of how any licensing is implemented. I think I can clarify that to better delineate that and make it clear any outside network requirement would see more scrutiny.

Reply:
I guess what I'm leaning towards here is rather than say "may be acceptable," we put something like "if you think you need this, talk to us first," without a whole lot of specific advice. If we start to get to the point where that approach isn't going to scale, we probably need to find some way to finance app reviews of such software to make it sustainable anyway, so we'll want to revise the policy when we get there. I think any process that involves humans manually reviewing for-profit apps for free is going to be unsustainable past a fairly small volume.

Reply:
That's fair. I probably am not expecting to get past a fairly small volume anytime soon, as it is though...